### PR TITLE
poseidon: better config

### PR DESCRIPTION
**Pull request by tcoratger**
## Summary
What does this PR change?

@Nashtare I think this way can help to add new configs and fields. I think in the future we want everything to be compile time generics but we need a bit more progress for this.

Fixes #<issue-id>

## Why
What problem does this solve? Why is this approach correct?

## Changes
- [ ] Key change 1
- [ ] Key change 2

## Acceptance criteria
- [ ] <explicit condition that proves "done">
- [ ] <explicit condition that proves "done">

## Tests
- [ ] Unit tests added/updated
- [ ] Existing tests pass
- [ ] Edge cases considered (briefly note any)

## Benchmarks / Performance (if relevant)
- [ ] Before/after numbers included
- [ ] Benchmark command or link to CI perf output
- [ ] No unexpected regressions in CPU / memory / proof size / compile time

## Docs (if relevant)
- [ ] Rustdoc updated
- [ ] Book/README updated
- [ ] Examples updated

## Risk
What could break?

## Review notes
Where should reviewers focus on? Any tricky parts / follow-ups?

### CODE REVIEW

**Review by Nashtare — COMMENTED** [review]
Overall I really like the changes, it's much neater and readable. I am however questioning the simplification around hasher/compression parameters which I do not think is the right direction.

File: circuit/src/ops/poseidon2_perm/config.rs [hunk 1]
```diff
@@ -11,114 +11,215 @@ use crate::CircuitBuilderError;
 use crate::builder::NpoLoweringContext;
 use crate::types::{ExprId, WitnessId};
 
-/// Poseidon2 configuration used as a stable operation key and parameter source.
+/// Identifies the base field for a Poseidon2 configuration.
+///
+/// Used internally for dispatching to concrete field-specific AIR types.
+/// When adding support for a new prime field, add a variant here.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd, Serialize, Deserialize)]
+pub enum Poseidon2FieldId {
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply circuit/src/ops/poseidon2_perm/config.rs]: Fixed

**Comment by Nashtare** [comment]
Rebased in #449.
